File: 4_Behavioral_Design_Patterns/4.06_Momento/Python_Snippet/Momento.py
# Originator
class TextEditor:
    def __init__(self, content):
        self._content = content

    def set_content(self, content):
        self._content = content

    def get_content(self):
        return self._content

    def save(self):
        return TextEditorMemento(self._content)

    def restore(self, memento):
        self._content = memento.get_content()


# Memento
class TextEditorMemento:
    def __init__(self, content):
        self._content = content

    def get_content(self):
        return self._content


# Caretaker
class History:
    def __init__(self):
        self._mementos = []

    def save(self, memento):
        self._mementos.append(memento)

    def undo(self):
        if self._mementos:
            return self._mementos.pop()
        return None


# Usage
if __name__ == "__main__":
    # Create originator (text editor)
    editor = TextEditor("Hello, world!")

    # Create caretaker (history)
    history = History()

    # Save initial state
    history.save(editor.save())

    # Perform some edits
    editor.set_content("Hello, ")
    history.save(editor.save())

    editor.set_content("Hello, Python!")
    history.save(editor.save())

    # Undo to previous state
    editor.restore(history.undo())
    print(editor.get_content())  # Output: Hello, 

    # Undo again
    editor.restore(history.undo())
    print(editor.get_content())  # Output: Hello, world!

File: 4_Behavioral_Design_Patterns/4.06_Momento/Java_Snippet/Momento.java
import java.util.Stack;

// Originator
class TextEditor {
    private String content;

    public TextEditor(String content) {
        this.content = content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }

    public TextEditorMemento save() {
        return new TextEditorMemento(content);
    }

    public void restore(TextEditorMemento memento) {
        this.content = memento.getContent();
    }
}

// Memento
class TextEditorMemento {
    private String content;

    public TextEditorMemento(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }
}

// Caretaker
class History {
    private Stack<TextEditorMemento> mementos = new Stack<>();

    public void save(TextEditorMemento memento) {
        mementos.push(memento);
    }

    public TextEditorMemento undo() {
        if (!mementos.isEmpty()) {
            return mementos.pop();
        }
        return null;
    }
}

// Usage
public class Momento {
    public static void main(String[] args) {
        // Create originator (text editor)
        TextEditor editor = new TextEditor("Hello, world!");

        // Create caretaker (history)
        History history = new History();

        // Save initial state
        history.save(editor.save());

        // Perform some edits
        editor.setContent("Hello, ");
        history.save(editor.save());

        editor.setContent("Hello, Java!");
        history.save(editor.save());

        // Undo to previous state
        editor.restore(history.undo());
        System.out.println(editor.getContent()); // Output: Hello,

        // Undo again
        editor.restore(history.undo());
        System.out.println(editor.getContent()); // Output: Hello, world!
    }
}
